File: Gillian-C/examples/act/bst.c
#include <stdlib.h>
#define FALSE 0
#define TRUE 1

typedef struct bstn {
    int value;
    struct bstn* left;
    struct bstn* right;
} BST;

BST* a_makeNode(int v) {
  BST* new_node = malloc(sizeof(BST));
  new_node->value = v;
  new_node->left = NULL;
  new_node->right = NULL;
  return new_node;
}


BST* b_insert(int v, BST* t) {
  BST* tmp;
  if (t == NULL) {
    return a_makeNode(v);
  };
  if (v < t->value) {
    tmp = b_insert(v, t->left);
    t->left = tmp;
  } else if (v > t->value) {
    tmp = b_insert(v, t->right);
    t->right = tmp;
  };
  return t;
}


int c_find (int v, BST* t) {
  if (t == NULL) {
    return FALSE;
  } else if (v == t->value) {
    return TRUE;
  } else if (v < t->value) {
    return c_find(v, t->left);
  } else { /* the only last case is v > t->value */
    return c_find(v, t->right);
  }
}

int d_find_min (BST* t) {
  if (t->left == NULL) {
    return t->value;
  } else {
    return d_find_min(t->left);
  }
}

BST* e_remove(int v, BST* t) {
  BST* ret_node;
  if (t == NULL) {
    return NULL;
  } else if (v == t->value) {
    if (t->left == NULL) {
      ret_node = t->right;
      free(t);
      return ret_node;
    } else if (t->right == NULL) {
      ret_node = t->left;
      free(t);
      return ret_node;
    } else {
      int min = d_find_min(t->right);
      t->right = e_remove(min, t->right);
      t->value = min;
    }
  } else if (v < t->value) {
    t->left = e_remove(v, t->left);
  } else {
    t->right = e_remove(v, t->right);
  };
  return t;
}

int peek_unsafe (int v, BST* t) {
  return t -> value;
}

int peek_left_unsafe (int v, BST* t) {
  return t -> left -> value;
}

int peek_right_unsafe (int v, BST* t) {
  return t -> right -> value;
}

File: Gillian-JS/Examples/Bi/CaseStudies/BST.js
'use strict';

/*
  @id a_make_node
*/
function a_make_node(v)
{
  /* Annotation */ isNumber(v);

  var node = {
    value : v,
    left  : null,
    right : null
  };
  return node;
}

/**
	@id b_insert
*/
function b_insert(v, t)
{

  /* Annotation */ isNumber(v);
  /* Annotation */ isNullableObject(t);

  if (t === null) {
  	return a_make_node(v);
  }

  /* Annotation */ isNumber(t.value);

  if (v < t.value) {
    t.left = b_insert(v, t.left);
  }
  else if (v > t.value) {
    t.right = b_insert(v, t.right);
  }

  return t;
}

/* @id c_find */
function c_find (v, t)
{
  /* Annotation */ isNumber(v);
  /* Annotation */ isNullableObject(t);

	if (t === null)
    return false;

  /* Annotation */ isNumber(t.value);

  if (v === t.value)
    return true;

  if (v < t.value)
    return c_find(v, t.left);

  return c_find(v, t.right);
}

/* @id d_findMin */
function d_findMin(t)
{
  /* Annotation */ isObject(t);
  /* Annotation */ isNumber(t.value);
  /* Annotation */ isNullableObject(t.left);

  if (t.left === null)
    return t.value;

	return d_findMin(t.left);
}

/* @id e_remove */
function e_remove(v, t)
{
  /* Annotation */ isNumber(v);
  /* Annotation */ isNullableObject(t);

	if (t === null)
    return null;

  /* Annotation */ isNumber(t.value);

	if (v === t.value) {
      /* Annotation */ isNullableObject(t.left);
		  if (t.left === null) {
				return t.right;
			}
    else
      /* Annotation */ isNullableObject(t.right);
	  	if (t.right === null) {
	  			return t.left;
			}
		else {
			var min = d_findMin(t.right);
			t.right = e_remove(min, t.right);
			t.value = min;
		}
	}
  else if (v < t.value) {
    /* Annotation */ isNullableObject(t.left);
    t.left = e_remove(v, t.left)
  }
  else {
    /* Annotation */ isNullableObject(t.left);
    t.right = e_remove(v, t.right)
  }

  return t;
}
